### PR TITLE
Grosmann/fpga hard macro map

### PR DESCRIPTION
**Pull request by petergrossmann21**
In this PR we add support for DSP and Block RAM mapping.

The DSP mapping is offered in three flavors -- conventional techmapping, an extraction pass, and blackboxing.  Which of these three is used is determined by a yosys tool/task variable called dsp_map_method, which defaults to "techmap" if the user does nothing.

To realize the above, an FPGA part with DSP blocks must implement one or more of these methods by specifying a corresponding yosys map file (Verilog) and any +define arguments that the verilog might require.  For techmapping, there are six required options to align the width of the DSP block so that mapping can be done on top of yosys's built-in multiplier mapping.  For the others, there are no requirements other than those specific to the FPGA part.

The BRAM mapping works exactly one way and requires specification of yosys's tool-specific memory map file format (plain text, see yosys documentation for details) and a Verilog memory techmap file that must be codesigned with the memory map file.  These files must be specified for an FPGA part that has block RAMs.

In all cases for both DSP and BRAM, failure to specify the required mapping files (verilog/bram text) causes the corresponding yosys mapping pass to be skipped.  This is a tradeoff in error checking that allows FPGAs that do not contain these blocks to omit all related settings in their part drivers.

### CODE REVIEW

File: examples/fpga_flow/designs/matrix_multiply/matrix_multiply.v [hunk 1]
```diff
@@ -0,0 +1,472 @@
+//matrix_multiply.v
+//Peter Grossmann
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/matrix_multiply.v:2]: add brief description of what this is doing

File: examples/fpga_flow/designs/matrix_multiply/matrix_multiply_control.v [hunk 2]
```diff
@@ -0,0 +1,163 @@
+//matrix_multiply_control.v
+//Peter Grossmann
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/matrix_multiply_control.v:2]: add brief description of what this is doing

File: examples/fpga_flow/designs/matrix_multiply/row_col_data_mux.v [hunk 3]
```diff
@@ -0,0 +1,46 @@
+//row_col_data_mux.v
+//Peter Grossmann
+//9 July 2023
+//$Id$
+//$Log$
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_data_mux.v:5]: remove

File: examples/fpga_flow/designs/matrix_multiply/row_col_data_mux.v [hunk 4]
```diff
@@ -0,0 +1,46 @@
+//row_col_data_mux.v
+//Peter Grossmann
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_data_mux.v:2]: add brief description of what this is doing

File: examples/fpga_flow/designs/matrix_multiply/row_col_memory.v [hunk 5]
```diff
@@ -0,0 +1,110 @@
+//row_col_memory.v
+//Peter Grossmann
+//6 July 2023
+//$Id$
+//$Log$
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_memory.v:5]: remove

File: examples/fpga_flow/designs/matrix_multiply/row_col_memory.v [hunk 6]
```diff
@@ -0,0 +1,110 @@
+//row_col_memory.v
+//Peter Grossmann
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_memory.v:2]: add brief description of what this is doing

File: examples/fpga_flow/designs/matrix_multiply/row_col_multiply.v [hunk 7]
```diff
@@ -0,0 +1,30 @@
+//row_col_multiply.v
+//Peter Grossmann
+//6 July 2023
+//$Id$
+//$Log$
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_multiply.v:5]: remove

File: examples/fpga_flow/designs/matrix_multiply/row_col_multiply.v [hunk 8]
```diff
@@ -0,0 +1,30 @@
+//row_col_multiply.v
+//Peter Grossmann
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_multiply.v:2]: add brief description of what this is doing

File: examples/fpga_flow/designs/matrix_multiply/row_col_product_adder.v [hunk 9]
```diff
@@ -0,0 +1,69 @@
+//row_col_product_adder.v
+//Peter Grossmann
+//6 July 2023
+//$Id$
+//$Log$
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_product_adder.v:5]: remove

File: examples/fpga_flow/designs/matrix_multiply/row_col_product_adder.v [hunk 10]
```diff
@@ -0,0 +1,69 @@
+//row_col_product_adder.v
+//Peter Grossmann
```
Comment (gadfort) [inline examples/fpga_flow/designs/matrix_multiply/row_col_product_adder.v:2]: add brief description of what this is doing

File: examples/fpga_flow/techlib/example_arch_bram_memory_map.txt [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+	byte 8;
+        widths 1 2 4 8 per_port;
+        cost 3;
+        init none;
+        port srsw "A" "B" {
+                clock posedge;
+                rden;
+        }
+
+}
```
Comment (gadfort) [inline examples/fpga_flow/techlib/example_arch_bram_memory_map.txt:12]: could you add a set of `# comments` to describe the type of memory getting mapped here

File: examples/fpga_flow/techlib/example_arch_techmap_dsp_extract.v [hunk 12]
```diff
@@ -0,0 +1,18 @@
+
+`ifdef BLACKBOX_MACROS
+(* blackbox *)
+`else
+`endif(* extract_order=5 *)
```
Comment (gadfort) [inline examples/fpga_flow/techlib/example_arch_techmap_dsp_extract.v:5]: what does this do?

File: siliconcompiler/fpgas/vpr_example.py [hunk 13]
```diff
@@ -69,6 +70,30 @@ def setup(chip):
...
+            fpga.add('fpga', part_name, 'file', 'yosys_dsp_techmap',
+                     os.path.join(techlib_root, 'example_arch_techmap_dsp.v'))
+
+            fpga.add('fpga', part_name, 'file', 'yosys_dsp_extractlib',
+                     os.path.join(techlib_root, 'example_arch_techmap_dsp_extract.v'))
+
+            # This techlib combines the blackboxing and extraction in one library
+            # with a +define; build it out accordingly
+            fpga.add('fpga', part_name, 'file', 'yosys_dsp_blackboxlib',
+                     os.path.join(techlib_root, 'example_arch_techmap_dsp_extract.v'))
```
Comment (gadfort) [inline siliconcompiler/fpgas/vpr_example.py:82]: why are two needed?

File: siliconcompiler/fpgas/vpr_example.py [hunk 14]
```diff
@@ -69,6 +70,30 @@ def setup(chip):
...
+
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_A_MAXWIDTH=18')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_B_MAXWIDTH=18')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_A_MINWIDTH=2')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_B_MINWIDTH=2')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_NAME=_dsp_block_')
+
+            fpga.add('fpga', part_name, 'var', 'dsp_blackbox_options', 'BLACKBOX_MACROS')
```
Comment (gadfort) [inline siliconcompiler/fpgas/vpr_example.py:90]: why is this needed?

File: siliconcompiler/fpgas/vpr_example.py [hunk 15]
```diff
@@ -69,6 +70,30 @@ def setup(chip):
...
+            # This techlib combines the blackboxing and extraction in one library
+            # with a +define; build it out accordingly
+            fpga.add('fpga', part_name, 'file', 'yosys_dsp_blackboxlib',
+                     os.path.join(techlib_root, 'example_arch_techmap_dsp_extract.v'))
+
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_A_MAXWIDTH=18')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_B_MAXWIDTH=18')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_A_MINWIDTH=2')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_B_MINWIDTH=2')
+            fpga.add('fpga', part_name, 'var', 'dsp_options', 'DSP_NAME=_dsp_block_')
```
Comment (gadfort) [inline siliconcompiler/fpgas/vpr_example.py:88]: rename to `yosys_dsp_options`

**Review by gadfort — CHANGES_REQUESTED** [review]

File: siliconcompiler/fpgas/vpr_example.py (hunk 13 above)
Comment (petergrossmann21) [reply siliconcompiler/fpgas/vpr_example.py:82]: In the new implementation, this will be needed to trigger the file to be read both at the read_verilog step and the extract step; it's not difficult to imagine a setup where you would want to use a given file for only one and not the other, but in this example one file is suitable for both.

**Review by gadfort — CHANGES_REQUESTED** [review]

File: siliconcompiler/tools/yosys/syn_fpga.py [hunk 16]
```diff
@@ -38,6 +38,18 @@ def setup_fpga(chip):
              ",".join(['fpga', part_name, 'lutsize']),
              step=step, index=index)
 
+    # Verify memory techmapping setup.  If a memory libmap
+    # is provided a memory techmap verilog file is needed too
+    if chip.valid('fpga', part_name, 'file', 'yosys_memory_libmap') and \
+       chip.get('fpga', part_name, 'file', 'yosys_memory_libmap'):
+
```
[the remark this answers is not in the document]
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.py]: Done
